File: packages/fsm-async-worker-sdk-go/actor_worker.go
// Package asyncworkersdk is the Go worker SDK for the pgfsm Activity Gateway.
//
// A worker process built on it connects to the gateway's sidecar Unix socket,
// registers a set of actors, and serves the invocations the gateway routes to
// them over the pgfsm.sidecargateway.v1.SidecarGatewayService gRPC stream
// (stubs from the pgfsm-proto-codegen Go module). It never opens a database
// connection — that stays in the gateway.
//
// You normally don't write against this package directly: @pgfsm/compiler's
// generate-async-logic writes a small main.go that maps the project's
// generated actor registry into [ActorRegistration]s and calls
// [RunActorWorkerCLI].
//
// Moved here from fsm-compiler-ts's go/worker-sdk-sdk.eta (#370), which used
// to write this whole file into every project as async-worker/go/sdk.go.
//
// Unlike the TypeScript and Python SDKs there's no dynamic loading step: Go
// has no practical runtime mechanism to load a function out of a .go source
// file the way import() or importlib can (Go plugins need exact toolchain
// matching between plugin and host). ActorWorker takes an explicit
// []ActorRegistration built by the binary that uses it, so the actor functions
// are linked into that binary and a missing or mistyped actor fails the build,
// not worker startup.
//
// grpc-go's client stream gives a synchronous-looking Send/Recv pair backed by
// goroutines internally, so unlike the other SDKs there's no push-queue: only
// a mutex around Send (grpc-go client streams don't allow concurrent Send
// calls; Recv is only ever called from Run's goroutine).
package asyncworkersdk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"sync"
	"sync/atomic"
	"time"

	sidecargatewayv1 "github.com/pgfsm/fsm/packages/fsm-proto-codegen/gen/go/sidecargateway/v1"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// DefaultHeartbeatMs is the default heartbeat interval, in milliseconds.
const DefaultHeartbeatMs = 5000

// RegisteredActor is the generated pgfsm.sidecargateway.v1.RegisteredActor
// message: an actor's identity as the gateway sees it.
type RegisteredActor = sidecargatewayv1.RegisteredActor

// ActorHandler is an actor's compiled-in implementation. A returned error is
// reported to the gateway as an INTERNAL invoke error; a panic is recovered
// and reported the same way rather than crashing the worker process.
type ActorHandler func(input any) (any, error)

// ActorRegistration is one actor to register with the gateway: its identity
// (Meta) and its handler. Meta is a pointer because generated protobuf
// messages must not be copied.
type ActorRegistration struct {
	Meta    *RegisteredActor
	Handler ActorHandler
}

// NewActorRegistration builds a registration from the six identity fields the
// compiler's generated registry carries, plus the handler.
func NewActorRegistration(
	parentFsmName, parentFsmVersion, asyncOperationType, asyncOperationName,
	asyncOperationVersion, asyncOperationLanguage string,
	handler ActorHandler,
) ActorRegistration {
	return ActorRegistration{
		Meta: &RegisteredActor{
			ParentFsmName:          parentFsmName,
			ParentFsmVersion:       parentFsmVersion,
			AsyncOperationType:     asyncOperationType,
			AsyncOperationName:     asyncOperationName,
			AsyncOperationVersion:  asyncOperationVersion,
			AsyncOperationLanguage: asyncOperationLanguage,
		},
		Handler: handler,
	}
}

// ActorWorkerOptions configures an [ActorWorker].
type ActorWorkerOptions struct {
	WorkerID          string
	GatewaySocketPath string
	// HeartbeatMs is the heartbeat interval; zero means DefaultHeartbeatMs.
	HeartbeatMs int
}

// ActorKey is the key the gateway's invoke fields are matched against: all six
// identity fields joined with "@", same as the TypeScript/Python/Rust SDKs.
func ActorKey(parentFsmName, parentFsmVersion, asyncOperationType, asyncOperationName, asyncOperationVersion, asyncOperationLanguage string) string {
	return fmt.Sprintf("%s@%s@%s@%s@%s@%s", parentFsmName, parentFsmVersion, asyncOperationType, asyncOperationName, asyncOperationVersion, asyncOperationLanguage)
}

// ActorWorker registers a set of actors with the gateway and serves their
// invocations. Create one with [NewActorWorker].
type ActorWorker struct {
	options    ActorWorkerOptions
	handlers   map[string]ActorHandler
	registered []*RegisteredActor
	logger     *slog.Logger

	stopped atomic.Bool
	// mu guards stream and every Send on it.
	mu     sync.Mutex
	stream sidecargatewayv1.SidecarGatewayService_SessionClient
}

// NewActorWorker returns a worker for registrations. Nothing connects until
// [ActorWorker.Run].
func NewActorWorker(options ActorWorkerOptions, registrations []ActorRegistration) *ActorWorker {
	if options.HeartbeatMs <= 0 {
		options.HeartbeatMs = DefaultHeartbeatMs
	}
	handlers := make(map[string]ActorHandler, len(registrations))
	registered := make([]*RegisteredActor, 0, len(registrations))
	for _, reg := range registrations {
		m := reg.Meta
		handlers[ActorKey(m.GetParentFsmName(), m.GetParentFsmVersion(), m.GetAsyncOperationType(), m.GetAsyncOperationName(), m.GetAsyncOperationVersion(), m.GetAsyncOperationLanguage())] = reg.Handler
		registered = append(registered, m)
	}
	return &ActorWorker{
		options:    options,
		handlers:   handlers,
		registered: registered,
		logger:     slog.Default().With("component", "pgfsm.async_worker_sdk"),
	}
}

// RegisteredActors returns the actors this worker registers, in order.
func (w *ActorWorker) RegisteredActors() []*RegisteredActor {
	return w.registered
}

// Run registers every actor and serves invocations until [ActorWorker.Stop] is
// called or the gateway ends the stream, then closes its connection. A
// shutdown via Stop returns nil.
func (w *ActorWorker) Run() error {
	if len(w.registered) == 0 {
		return errors.New("no actors to register, refusing to start worker")
	}
	if w.stopped.Load() {
		return nil
	}

	conn, err := grpc.NewClient("unix://"+w.options.GatewaySocketPath, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return err
	}
	defer conn.Close()

	stream, err := sidecargatewayv1.NewSidecarGatewayServiceClient(conn).Session(context.Background())
	if err != nil {
		return err
	}
	w.mu.Lock()
	w.stream = stream
	w.mu.Unlock()

	if err := w.send(&sidecargatewayv1.SessionRequest{
		Payload: &sidecargatewayv1.SessionRequest_Register{
			Register: &sidecargatewayv1.Register{
				WorkerId:        w.options.WorkerID,
				Language:        "go",
				ProtocolVersion: "1.0",
				Actors:          w.registered,
			},
		},
	}); err != nil {
		return fmt.Errorf("sending register: %w", err)
	}

	first, err := stream.Recv()
	if err != nil {
		return fmt.Errorf("waiting for register_ack: %w", err)
	}
	ack, ok := first.GetPayload().(*sidecargatewayv1.SessionResponse_RegisterAck)
	if !ok {
		return fmt.Errorf("expected register_ack but got %T", first.GetPayload())
	}
	if !ack.RegisterAck.GetAccepted() {
		w.stopped.Store(true)
		w.closeSend()
		return errors.New("gateway rejected registration")
	}

	w.logger.Info("worker registered actors with the gateway", "worker_id", w.options.WorkerID, "actors", len(w.registered))

	heartbeatDone := make(chan struct{})
	stopHeartbeat := make(chan struct{})
	go func() {
		defer close(heartbeatDone)
		ticker := time.NewTicker(time.Duration(w.options.HeartbeatMs) * time.Millisecond)
		defer ticker.Stop()
		for {
			select {
			case <-stopHeartbeat:
				return
			case <-ticker.C:
				if w.stopped.Load() {
					return
				}
				_ = w.send(&sidecargatewayv1.SessionRequest{
					Payload: &sidecargatewayv1.SessionRequest_Heartbeat{
						Heartbeat: &sidecargatewayv1.Heartbeat{WorkerId: w.options.WorkerID},
					},
				})
			}
		}
	}()

	err = w.serveLoop(stream)
	w.stopped.Store(true)
	close(stopHeartbeat)
	<-heartbeatDone
	w.closeSend()
	return err
}

// Stop asks the gateway to unregister this worker and ends the session. Safe
// to call from any goroutine (e.g. a signal handler), more than once, and
// before Run has connected.
func (w *ActorWorker) Stop() {
	if w.stopped.Swap(true) {
		return
	}
	_ = w.send(&sidecargatewayv1.SessionRequest{
		Payload: &sidecargatewayv1.SessionRequest_Unregister{
			Unregister: &sidecargatewayv1.Unregister{WorkerId: w.options.WorkerID},
		},
	})
	// Half-close rather than closing the connection: the gateway sees the
	// unregister and end of stream, ends its side, and Run returns cleanly.
	w.closeSend()
}

func (w *ActorWorker) send(req *sidecargatewayv1.SessionRequest) error {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.stream == nil {
		return errors.New("not connected")
	}
	return w.stream.Send(req)
}

func (w *ActorWorker) closeSend() {
	w.mu.Lock()
	defer w.mu.Unlock()
	if w.stream != nil {
		_ = w.stream.CloseSend()
	}
}

func (w *ActorWorker) serveLoop(stream sidecargatewayv1.SidecarGatewayService_SessionClient) error {
	for {
		resp, err := stream.Recv()
		if err != nil {
			// A clean end of stream, or any error after Stop, is a normal
			// shutdown.
			if errors.Is(err, io.EOF) || w.stopped.Load() {
				return nil
			}
			return err
		}
		switch p := resp.GetPayload().(type) {
		case *sidecargatewayv1.SessionResponse_Invoke:
			w.handleInvoke(p.Invoke)
		case *sidecargatewayv1.SessionResponse_Cancel:
			// Invokes run to completion; nothing to cancel.
		}
	}
}

func (w *ActorWorker) handleInvoke(body *sidecargatewayv1.Invoke) {
	key := ActorKey(body.GetParentFsmName(), body.GetParentFsmVersion(), body.GetAsyncOperationType(), body.GetAsyncOperationName(), body.GetAsyncOperationVersion(), body.GetAsyncOperationLanguage())
	handler, ok := w.handlers[key]
	if !ok {
		w.logger.Warn("invoke for unknown actor", "invoke_id", body.GetInvokeId(), "actor", key)
		w.sendError(body.GetInvokeId(), "NOT_FOUND", fmt.Sprintf("actor not found: %s", key))
		return
	}

	var input any
	if s := body.GetInputJson(); s != "" {
		if err := json.Unmarshal([]byte(s), &input); err != nil {
			w.sendError(body.GetInvokeId(), "INTERNAL", fmt.Sprintf("invalid input_json: %v", err))
			return
		}
	}

	started := time.Now()
	output, err := safeInvoke(handler, input)
	if err != nil {
		w.logger.Error("actor failed", "actor", key, "error", err)
		w.sendError(body.GetInvokeId(), "INTERNAL", err.Error())
		return
	}

	outputJSON, err := json.Marshal(output)
	if err != nil {
		w.sendError(body.GetInvokeId(), "INTERNAL", fmt.Sprintf("output is not JSON-serializable: %v", err))
		return
	}

	_ = w.send(&sidecargatewayv1.SessionRequest{
		Payload: &sidecargatewayv1.SessionRequest_InvokeResult{
			InvokeResult: &sidecargatewayv1.InvokeResult{
				InvokeId:   body.GetInvokeId(),
				OutputJson: string(outputJSON),
				DurationMs: uint32(time.Since(started).Milliseconds()),
			},
		},
	})
}

func safeInvoke(handler ActorHandler, input any) (output any, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic: %v", r)
		}
	}()
	return handler(input)
}

func (w *ActorWorker) sendError(invokeID, code, message string) {
	_ = w.send(&sidecargatewayv1.SessionRequest{
		Payload: &sidecargatewayv1.SessionRequest_InvokeError{
			InvokeError: &sidecargatewayv1.InvokeError{
				InvokeId: invokeID,
				Error: &sidecargatewayv1.InvokeErrorDetail{
					Code:      code,
					Message:   message,
					Retriable: false,
				},
			},
		},
	})
}

File: packages/fsm-async-worker-sdk-rust/src/actor_worker.rs
//! `ActorWorker` — moved here from fsm-compiler-ts's `rust/worker-sdk-sdk.eta`
//! (#368), which used to write this whole file into every project as
//! `async-worker/rust/src/sdk.rs`.
//!
//! Connects to the gateway's sidecar Unix socket via the generated
//! `pgfsm.sidecargateway.v1.SidecarGatewayService` bidi-streaming client (the
//! `pgfsm-proto-codegen` crate), registers a compiled-in actor registry, and
//! serves invoke requests.
//!
//! Unlike the TypeScript and Python SDKs, there's no dynamic loading step:
//! Rust has no runtime mechanism to load a function out of a `.rs` source file
//! the way `import()` or `importlib` can. `ActorWorker` takes an explicit
//! `Vec<ActorRegistration>` built by the binary that uses it (the generated
//! worker's `main.rs`), so the actor functions are linked into that binary and
//! "verification" happens at compile time.
//!
//! Outgoing messages (register, heartbeat, invoke_result, invoke_error,
//! unregister) are pushed onto a `tokio::sync::mpsc::UnboundedSender` whose
//! receiver, wrapped in an `UnboundedReceiverStream`, is tonic's request
//! stream — the Rust analogue of the TypeScript SDK's push-based AsyncQueue
//! and the Python SDK's queue-backed generator. Dropping the sender ends the
//! stream, which is how `stop()` (a plain sync method, safe to call from a
//! signal handler) shuts things down without an async context.

use pgfsm_proto_codegen::pgfsm::sidecargateway::v1::sidecar_gateway_service_client::SidecarGatewayServiceClient;
use pgfsm_proto_codegen::pgfsm::sidecargateway::v1::{
    session_request, session_response, Heartbeat, Invoke, InvokeError, InvokeErrorDetail,
    InvokeResult, Register, RegisteredActor, SessionRequest, SessionResponse, Unregister,
};
use serde_json::Value;
use std::collections::HashMap;
use std::panic::{catch_unwind, AssertUnwindSafe};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};
use tokio::sync::mpsc;
use tokio_stream::wrappers::UnboundedReceiverStream;
use tokio_stream::StreamExt;
use tonic::codec::Streaming;

/// Default heartbeat interval, in milliseconds.
pub const DEFAULT_HEARTBEAT_MS: u64 = 5000;

/// Error type returned by [`ActorWorker::run`].
pub type BoxError = Box<dyn std::error::Error + Send + Sync>;

/// An actor's compiled-in implementation. Panics are caught by the SDK and
/// reported to the gateway as an `INTERNAL` invoke error rather than crashing
/// the worker process — the Rust equivalent of a thrown exception in the
/// TypeScript/Python SDKs.
pub type ActorHandler = Box<dyn Fn(Value) -> Value + Send + Sync>;

/// One actor to register with the gateway: its identity (`meta`, the same
/// `RegisteredActor` message the gateway receives) and its handler.
pub struct ActorRegistration {
    pub meta: RegisteredActor,
    pub handler: ActorHandler,
}

impl ActorRegistration {
    /// Builds a registration from the six identity fields the compiler's
    /// generated registry carries, plus the handler.
    pub fn new(
        parent_fsm_name: &str,
        parent_fsm_version: &str,
        async_operation_type: &str,
        async_operation_name: &str,
        async_operation_version: &str,
        async_operation_language: &str,
        handler: impl Fn(Value) -> Value + Send + Sync + 'static,
    ) -> Self {
        Self {
            meta: RegisteredActor {
                parent_fsm_name: parent_fsm_name.to_string(),
                parent_fsm_version: parent_fsm_version.to_string(),
                async_operation_type: async_operation_type.to_string(),
                async_operation_name: async_operation_name.to_string(),
                async_operation_version: async_operation_version.to_string(),
                async_operation_language: async_operation_language.to_string(),
                ..Default::default()
            },
            handler: Box::new(handler),
        }
    }
}

pub struct ActorWorkerOptions {
    pub worker_id: String,
    pub gateway_socket_path: String,
    pub heartbeat_ms: u64,
}

/// The key the gateway's invoke fields are matched against: all six identity
/// fields joined with `@`, same as the TypeScript and Python SDKs.
pub fn actor_key(
    parent_fsm_name: &str,
    parent_fsm_version: &str,
    async_operation_type: &str,
    async_operation_name: &str,
    async_operation_version: &str,
    async_operation_language: &str,
) -> String {
    format!(
        "{}@{}@{}@{}@{}@{}",
        parent_fsm_name,
        parent_fsm_version,
        async_operation_type,
        async_operation_name,
        async_operation_version,
        async_operation_language
    )
}

pub struct ActorWorker {
    options: ActorWorkerOptions,
    handlers: HashMap<String, ActorHandler>,
    registered: Vec<RegisteredActor>,
    stopped: Arc<AtomicBool>,
    outbox: Mutex<Option<mpsc::UnboundedSender<SessionRequest>>>,
}

impl ActorWorker {
    pub fn new(options: ActorWorkerOptions, registrations: Vec<ActorRegistration>) -> Self {
        let mut handlers = HashMap::new();
        let mut registered = Vec::with_capacity(registrations.len());
        for reg in registrations {
            let key = actor_key(
                &reg.meta.parent_fsm_name,
                &reg.meta.parent_fsm_version,
                &reg.meta.async_operation_type,
                &reg.meta.async_operation_name,
                &reg.meta.async_operation_version,
                &reg.meta.async_operation_language,
            );
            handlers.insert(key, reg.handler);
            registered.push(reg.meta);
        }
        Self {
            options,
            handlers,
            registered,
            stopped: Arc::new(AtomicBool::new(false)),
            outbox: Mutex::new(None),
        }
    }

    /// The actors this worker registers, in registration order.
    pub fn registered_actors(&self) -> &[RegisteredActor] {
        &self.registered
    }

    /// Registers every actor and serves invocations until [`stop`](Self::stop)
    /// is called or the gateway ends the stream. The gRPC channel is dropped
    /// (closed) before this returns.
    pub async fn run(&self) -> Result<(), BoxError> {
        if self.registered.is_empty() {
            return Err("no actors to register, refusing to start worker".into());
        }
        if self.stopped.load(Ordering::SeqCst) {
            return Ok(());
        }

        let endpoint = tonic::transport::Endpoint::from_shared(format!(
            "unix://{}",
            self.options.gateway_socket_path
        ))?;
        let channel = endpoint.connect().await?;
        let mut client = SidecarGatewayServiceClient::new(channel);

        let (tx, rx) = mpsc::unbounded_channel::<SessionRequest>();
        tx.send(SessionRequest {
            payload: Some(session_request::Payload::Register(Register {
                worker_id: self.options.worker_id.clone(),
                language: "rust".to_string(),
                protocol_version: "1.0".to_string(),
                actors: self.registered.clone(),
            })),
        })?;
        let heartbeat_tx = tx.clone();
        *self.outbox.lock().unwrap() = Some(tx);

        let outbound = UnboundedReceiverStream::new(rx);
        let response = client.session(outbound).await?;
        let mut inbound = response.into_inner();

        let first = inbound
            .next()
            .await
            .ok_or("expected register_ack but got EOF")??;
        let register_ack = match first.payload {
            Some(session_response::Payload::RegisterAck(ack)) => ack,
            other => return Err(format!("expected register_ack but got {:?}", other).into()),
        };
        if !register_ack.accepted {
            self.stopped.store(true, Ordering::SeqCst);
            self.close_outbox(None);
            return Err("gateway rejected registration".into());
        }

        log::info!(
            "Worker {} registered {} actor(s) with the gateway",
            self.options.worker_id,
            self.registered.len()
        );

        let heartbeat_handle = {
            let stopped = self.stopped.clone();
            let worker_id = self.options.worker_id.clone();
            let heartbeat_ms = self.options.heartbeat_ms;
            tokio::spawn(async move {
                while !stopped.load(Ordering::SeqCst) {
                    tokio::time::sleep(Duration::from_millis(heartbeat_ms)).await;
                    if stopped.load(Ordering::SeqCst) {
                        break;
                    }
                    let msg = SessionRequest {
                        payload: Some(session_request::Payload::Heartbeat(Heartbeat {
                            worker_id: worker_id.clone(),
                        })),
                    };
                    if heartbeat_tx.send(msg).is_err() {
                        break;
                    }
                }
            })
        };

        let serve_result = self.serve_loop(&mut inbound).await;

        self.stopped.store(true, Ordering::SeqCst);
        heartbeat_handle.abort();
        self.close_outbox(None);

        serve_result
    }

    /// Asks the gateway to unregister this worker and ends the session. Safe to
    /// call from any thread (including a signal handler) and more than once.
    pub fn stop(&self) {
        if self.stopped.swap(true, Ordering::SeqCst) {
            return;
        }
        self.close_outbox(Some(SessionRequest {
            payload: Some(session_request::Payload::Unregister(Unregister {
                worker_id: self.options.worker_id.clone(),
            })),
        }));
    }

    /// Sends `final_message` (if any) and drops the stored sender — dropping
    /// the last `UnboundedSender` ends the `UnboundedReceiverStream` tonic is
    /// reading the request stream from, which is how a Rust client ends its
    /// side of a bidi call.
    fn close_outbox(&self, final_message: Option<SessionRequest>) {
        let mut guard = self.outbox.lock().unwrap();
        if let Some(tx) = guard.take() {
            if let Some(msg) = final_message {
                let _ = tx.send(msg);
            }
        }
    }

    fn push(&self, msg: SessionRequest) {
        if let Some(tx) = self.outbox.lock().unwrap().as_ref() {
            let _ = tx.send(msg);
        }
    }

    async fn serve_loop(&self, inbound: &mut Streaming<SessionResponse>) -> Result<(), BoxError> {
        while !self.stopped.load(Ordering::SeqCst) {
            let response = match inbound.next().await {
                Some(Ok(r)) => r,
                Some(Err(status)) => {
                    // After stop() the gateway may close the stream with a
                    // status rather than a clean EOF; that's a normal shutdown.
                    if self.stopped.load(Ordering::SeqCst) {
                        break;
                    }
                    return Err(Box::new(status));
                }
                None => break,
            };

            match response.payload {
                Some(session_response::Payload::Cancel(_)) => continue,
                Some(session_response::Payload::Invoke(invoke)) => self.handle_invoke(invoke),
                _ => continue,
            }
        }
        Ok(())
    }

    fn handle_invoke(&self, body: Invoke) {
        let key = actor_key(
            &body.parent_fsm_name,
            &body.parent_fsm_version,
            &body.async_operation_type,
            &body.async_operation_name,
            &body.async_operation_version,
            &body.async_operation_language,
        );

        let handler = match self.handlers.get(&key) {
            Some(h) => h,
            None => {
                log::warn!("Invoke {} for unknown actor {}", body.invoke_id, key);
                self.send_error(
                    &body.invoke_id,
                    "NOT_FOUND",
                    &format!("actor not found: {}", key),
                );
                return;
            }
        };

        let input: Value = if body.input_json.trim().is_empty() {
            Value::Null
        } else {
            match serde_json::from_str(&body.input_json) {
                Ok(v) => v,
                Err(e) => {
                    self.send_error(
                        &body.invoke_id,
                        "INTERNAL",
                        &format!("invalid input_json: {}", e),
                    );
                    return;
                }
            }
        };

        let started = Instant::now();
        match catch_unwind(AssertUnwindSafe(|| handler(input))) {
            Ok(output) => {
                let output_json =
                    serde_json::to_string(&output).unwrap_or_else(|_| "null".to_string());
                let duration_ms = started.elapsed().as_millis().min(u32::MAX as u128) as u32;
                self.push(SessionRequest {
                    payload: Some(session_request::Payload::InvokeResult(InvokeResult {
                        invoke_id: body.invoke_id,
                        output_json,
                        duration_ms,
                    })),
                });
            }
            Err(panic_payload) => {
                let message = panic_message(&panic_payload);
                log::error!("Actor {} panicked: {}", key, message);
                self.send_error(&body.invoke_id, "INTERNAL", &message);
            }
        }
    }

    fn send_error(&self, invoke_id: &str, code: &str, message: &str) {
        self.push(SessionRequest {
            payload: Some(session_request::Payload::InvokeError(InvokeError {
                invoke_id: invoke_id.to_string(),
                error: Some(InvokeErrorDetail {
                    code: code.to_string(),
                    message: message.to_string(),
                    retriable: false,
                }),
                duration_ms: 0,
            })),
        });
    }
}

fn panic_message(payload: &Box<dyn std::any::Any + Send>) -> String {
    if let Some(s) = payload.downcast_ref::<&str>() {
        s.to_string()
    } else if let Some(s) = payload.downcast_ref::<String>() {
        s.clone()
    } else {
        "actor panicked".to_string()
    }
}
